core/src/raw/oio/write/block_write.rs

```diff
@@ -0,0 +1,275 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::pin::Pin;
+use std::sync::Arc;
+use std::task::ready;
+use std::task::Context;
+use std::task::Poll;
+
+use async_trait::async_trait;
+use futures::Future;
+use futures::FutureExt;
+use futures::StreamExt;
+
+use crate::raw::*;
+use crate::*;
+
+/// BlockWrite is used to implement [`Write`] based on block
+/// uploads. By implementing BlockWrite, services don't need to
+/// care about the details of uploading blocks.
+///
+/// # Architecture
+///
+/// The architecture after adopting [`BlockWrite`]:
+///
+/// - Services impl `BlockWrite`
+/// - `BlockWriter` impl `Write`
+/// - Expose `BlockWriter` as `Accessor::Writer`
+///
+/// # Notes
+///
+/// `BlockWrite` has an oneshot optimization when `write` has been called only once:
+///
+/// ```no_build
+/// w.write(bs).await?;
+/// w.close().await?;
+/// ```
+///
+/// We will use `write_once` instead of starting a new block upload.
+#[cfg_attr(not(target_arch = "wasm32"), async_trait)]
+#[cfg_attr(target_arch = "wasm32", async_trait(?Send))]
+pub trait BlockWrite: Send + Sync + Unpin + 'static {
```

Comment:
Please don't include commits that not belong to this PR.

Reply:
fixed